### ISSUE TITLE
Failed to update object metadata with AWS S3

### ISSUE THREAD

**Issue by timuralp** [buildbarn/bb-storage#93]
Following the change to unconditionally call `blob.Copy` in FindMissing, objects bigger than 5GB result in an error from AWS (https://github.com/buildbarn/bb-storage/pull/82). This is because they require a multipart upload and to update metadata on a multipart upload multiple calls have to be made:
1. initiate multipart upload
2. upload part copy with the same source
3. complete multipart upload with the new metadata

The limitation is documented here: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html. Specifically:
> You create a copy of your object up to 5 GB in size in a single atomic operation using this API. However, to copy an object greater than 5 GB, you must use the multipart upload Upload Part - Copy API.

This also means that objects > 5GB will require multiple calls on every pass through FindMissing (since the x-amz-copy-source-if-unmodified-since has to be passed to every UploadPartCopy call -- https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html).

Lastly, this needs to be fixed in gocloud, but currently, bb-storage will not work as intended for objects > 5GB.

The error will appear as follows from AWS:
```
---[ REQUEST POST-SIGN ]-----------------------------
PUT /junk HTTP/1.1
Host: 1space-test.s3.amazonaws.com
User-Agent: aws-sdk-go/1.34.5 (go1.14.4; linux; amd64)
Content-Length: 0
Authorization: AWS4-HMAC-SHA256 Credential=<key>/20200902/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-unmodified-since;x-amz-date;x-amz-meta-used;x-amz-metadata-directive, Signature=5f21ff6a41f0aae38c1a562d7d43f195deb0d572755fbc8ab93151147309fee4
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Copy-Source: <object>
X-Amz-Copy-Source-If-Unmodified-Since: Wed, 02 Sep 2020 16:39:02 GMT
X-Amz-Date: 20200902T163902Z
X-Amz-Meta-Used: 2020-09-02 16:39:02.003101746 +0000 UTC
X-Amz-Metadata-Directive: REPLACE
Accept-Encoding: gzip


-----------------------------------------------------
2020/09/02 09:39:02 DEBUG: Response s3/CopyObject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Wed, 02 Sep 2020 16:39:02 GMT
Server: AmazonS3
X-Amz-Id-2: uYk9/YTJ16he14eMk1DslODt1TrksjLk/1burYQ+gMo9gOZgeSCXhvUiT3qv+5Bx2Y1/vvfSU1Q=
X-Amz-Request-Id: E871C91CA74390BD


-----------------------------------------------------
2020/09/02 09:39:02 <Error><Code>InvalidRequest</Code><Message>The specified copy source is larger than the maximum allowable size for a copy source: 5368709120</Message><RequestId>E871C91CA74390BD</RequestId><HostId>uYk9/YTJ16he14eMk1DslODt1TrksjLk/1burYQ+gMo9gOZgeSCXhvUiT3qv+5Bx2Y1/vvfSU1Q=</HostId></Error>
ret: blob (key "<key> -> <key>") (code=Unknown): InvalidRequest: The specified copy source is larger than the maximum allowable size for a copy source: 5368709120
        status code: 400, request id: E871C91CA74390BD, host id: uYk9/YTJ16he14eMk1DslODt1TrksjLk/1burYQ+gMo9gOZgeSCXhvUiT3qv+5Bx2Y1/vvfSU1Q=
```

**Comment by timuralp** [comment]
@asuffield tagging you on this bug, because it's related to the FindMissing change in the merged PR.

I again wonder if the unconditional server side copy should be optional, as the penalty for MPU blobs is not insignificant (HEAD vs CreateMPU, at least two UploadPartCopy requests, CompleteMPU).

**Comment by asuffield** [comment]
Interesting tangle. There isn't an obviously correct answer here - adding an option to hide the error doesn't sound like it would help anybody, we still wouldn't be refreshing the expiry time.

It's tricky to fix in gocloud because it doesn't know the size of the object. There isn't enough of an API in gocloud for us to detect this and work around it.

**Comment by timuralp** [comment]
This is going to be blobstore-dependent, as well, and the unified interface in gocloud makes it a little bit more difficult. As a gocloud consumer, I would expect it to catch the error and switch to the Initiate MPU/UploadPart Copy/Complete MPU flow.

I think this could be fixed here: https://github.com/google/go-cloud/blob/master/blob/s3blob/s3blob.go#L689. The wrapper should catch the 400, interpret the reason, and switch to Create MPU (with any object metadata -- I don't think you can use the metadata replace directive here), UploadPartCopy with the options (e.g. x-amz-copy-source-if-unmodified-since) specified (optionally, figuring out the part boundaries or defaulting to whatever the client has configured), and then CompleteMultipart.

Note that if the default part boundaries are used, that's 5MB parts, so it would take `size/5MB + 2` total requests to `touch` the object.

**Comment by EdSchouten** [comment]
Could we file a bug against gocloud for this?

**Comment by timuralp** [comment]
There is a gocloud issue: https://github.com/google/go-cloud/issues/2711, blocked by a change in AWS Go SDK.

@asuffield @EdSchouten but it doesn't address the broader point of how you envision this interacting with multipart uploads. Specifically, the SDK change will use maximum part sizes for the copy requests, which means that parts might be re-sized (which I suspect is not free). The x-amz-copy-if-unmodified-since also applies to the _parts_, not the resulting MPU, so depending on how that's resolved, that's another issue. And lastly, again, this will cause two requests for create/complete multipart, as well as the intermediate uploadPartCopy requests, which is a bit more expensive.

I still believe this should be a behavior that someone could opt out of -- for instance, we would want to opt out of it. We already have a bb-storage fork, but I would prefer to reduce the need for it, rather than increase it.

**Comment by EdSchouten** [comment]
Just to clarify my position on it: I think that using CloudBlobAccess to store the CAS is a *very* bad idea. Here is a document I wrote some time ago that hopefully explains sufficiently why that is:

https://github.com/buildbarn/bb-adrs/blob/master/0002-storage.md

I think CloudBlobAccess (and RedisBlobAccess) are only good enough to store AC entries. I personally wouldn't use them for anything else. I wouldn't mind kicking it out of the tree entirely, but I know that would upset certain users.

Because of that, I don't really care about what happens to that backend's code. I'm willing to keep it in the tree, as long as it's not in the way in terms of development and its ugliness doesn't spill into the rest of the codebase. Think of it like the Linux kernel: CloudBlobAccess is not much more valuable to me than some obscure ISA card driver.

So how I envision how all of this interacts with multipart uploads: I honestly don't know. Please consider working together with folks like @asuffield who also use CloudBlobAccess. Maybe it's time that some of you collaborate on an ADR to describe your vision on all of this?

**Comment by EdSchouten** [comment]
Closing this, as CloudBlobAccess has been removed from the tree in commit 0efcb2afc433a6c1ec705a5f7c9130b58b98ed18.
